### PR TITLE
Denis/azure/terraform

### PR DESCRIPTION
**Pull request by DenisMishin**
Initial set of changes to support multi-cloud provisioning

Breaking changes:
1. All Terraform parameters must come from config files, no more hard-coded params in .tf templates 
2. `install_url` must have schema - i.e. s3:// for S3 files

### CODE REVIEW

File: e2e/framework/test_context.go [hunk 1]
```diff
@@ -149,11 +150,13 @@ type TestContextType struct {
 	// Wizard specifies whether wizard was used to bootstrap cluster
 	Wizard bool `json:"-" yaml:"-"`
 	// Provisioner defines the type of provisioner to use
-	Provisioner provisionerType `json:"provisioner" yaml:"provisioner" env:"ROBO_PROVISIONER"`
+	Provisioner provisionerType `json:"provisioner" yaml:"provisioner" `
+	// ProvisionTo defines cloud to deploy
+	ProvisionTo string `json:"provision_to" yaml:"provision_to" validate:"omitempty,eq=aws|eq=azure"`
```
Comment (r0mant) [inline e2e/framework/test_context.go]: I would name it `CloudProvider` or something similar to be consistent with the terminology we use in the rest of the codebase (e.g. in gravity).

File: e2e/framework/test_context.go [hunk 2]
```diff
@@ -149,11 +150,13 @@ type TestContextType struct {
...
 	// DumpCore specifies a command to collect all installation/operation logs
 	DumpCore bool `json:"-" yaml:"-"`
 	// StateDir specifies the location for test-specific temporary data
-	StateDir string `json:"state_dir" yaml:"state_dir" env:"ROBO_STATE_DIR"`
```
Comment (r0mant) [inline e2e/framework/test_context.go]: Why are we getting rid of ability to override parameters via env vars btw?

File: e2e/framework/test_context.go [hunk 3]
```diff
@@ -162,60 +165,63 @@ type TestContextType struct {
...
 	// Bandwagon defines the test configuration for post-install setup in bandwagon
 	Bandwagon BandwagonConfig `json:"bandwagon" yaml:"bandwagon"`
 	// WebDriverURL specifies optional WebDriver URL to use
-	WebDriverURL string `json:"web_driver_url,omitempty" yaml:"web_driver_url,omitempty" env:"ROBO_WEB_DRIVER_URL"`
+	WebDriverURL string `json:"web_driver_url,omitempty" yaml:"web_driver_url,omitempty" `
 	// Extensions groups arbitrary test step configuration
 	Extensions Extensions `json:"extensions,omitempty" yaml:"extensions,omitempty"`
 }
 
 type BandwagonConfig struct {
```
Comment (r0mant) [inline e2e/framework/test_context.go]: Godocs on all exported structs/fields (below too).

File: e2e/framework/test_context.go [hunk 4]
```diff
@@ -375,7 +351,22 @@ func newContextConfig(configFile string) error {
...
+	if err == nil {
+		return nil
+	}
+
+	log.Errorf("Configuration file %s has errors", configFile)
+	for _, fieldError := range err.(validator.ValidationErrors) {
+		log.Errorf("   Field %s=%v fails rule %s", fieldError.Field(), fieldError.Value(), fieldError.Tag())
+	}
+
+	return trace.Errorf("Configuration file fails validation test")
```
Comment (r0mant) [inline e2e/framework/test_context.go]: Lower-case first letter of error messages, for better embedding: https://github.com/golang/go/wiki/CodeReviewComments#error-strings (above too).

File: infra/config.go [hunk 5]
```diff
@@ -34,3 +34,63 @@ type StateNode struct {
 	// KeyPath defines the location of the SSH key
 	KeyPath string `json:"key_path"`
 }
+
+// AWSConfig describes AWS EC2 test configuration
+type AWSConfig struct {
+	AccessKey string `json:"access_key" yaml:"access_key" validate:"required"`
```
Comment (r0mant) [inline infra/config.go]: Godocs on these two.

File: infra/config.go [hunk 6]
```diff
@@ -34,3 +34,63 @@ type StateNode struct {
...
+// FIXME : replace with embedded validation rules
+func (r *AWSConfig) IsEmpty() bool {
+	return r.AccessKey == "" && r.SecretKey == ""
+}
+
+// AzureConfig specifies Azure cloud specific parameters
+// see https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-create-service-principal-portal
+// to obtain subscription_id, client_id, client_secret and tenant_id
+type AzureConfig struct {
+	SubscriptionId string `json:"subscription_id" yaml:"subscription_id" validate:"required"`
```
Comment (r0mant) [inline infra/config.go]: Godocs.

File: infra/terraform/file_transfer.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    otherwise robotest completes but node is unsusable as there's still background process running
+
+ 2. waiting for boostrap script to complete (or maybe replacing its behind-the-scenes with more explicit preparation here)
+
+ 3. being sure which command exactly failed and avoiding monstrous && style pipelining
+
+ 4. support for SCP for local tarballs
+*/
+
+// TransferFile takes file URL which may be S3 or HTTP or local file and transfers it to the machine
```
Comment (r0mant) [inline infra/terraform/file_transfer.go]: The comment seems to be misplaced.

File: infra/terraform/file_transfer.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+	}
+
+	cmd := fmt.Sprintf(`echo Testing if bootstrap completed && test -f /var/lib/bootstrap_complete && \
+		echo Cleaning up && rm -rf %[1]s/installer/* 
+		echo Downloading installer %[5]s ... && %[2]s && \
+		echo Creating installer dir && mkdir -p %[1]s/installer && \
+		echo Unpacking installer && tar -xvf %[3]s -C %[1]s/installer && \
+		echo Launching command %[4]s && cd %[1]s/installer && %[4]s`, homeDir, fetchCmd, outFile, command, fileUrl)
+
+	// cmd = fmt.Sprintf(`echo Launching command %[4]s && cd %[1]s/installer && %[4]s`,
```
Comment (r0mant) [inline infra/terraform/file_transfer.go]: Debugging remains?

File: infra/terraform/file_transfer.go [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+	var fetchCmd string
+	var homeDir = fmt.Sprintf("/home/%s", t.sshUser)
+	var outFile string
+	if strings.HasSuffix(fileUrl, ".tar.gz") {
+		outFile = fmt.Sprintf("%s/installer.tar.gz", homeDir)
+	} else if strings.HasSuffix(fileUrl, ".tar") {
+		outFile = fmt.Sprintf("%s/installer.tar", homeDir)
+	} else {
+		return "", trace.Errorf("Unsupported installer packaging %s", fileUrl)
```
Comment (r0mant) [inline infra/terraform/file_transfer.go]: That's pretty cryptic error message, something like "only .tar and .tar.gz are supported, got: .blah" would be clearer I think. Also, lowecase (as mentioned above).

File: infra/terraform/file_transfer.go [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+// fileUrl - file to download, could be S3:// or http(s)://
+// command - to run out of installer package
+func (t *terraform) makeRemoteCommand(fileUrl, command string) (string, error) {
+	u, err := url.Parse(fileUrl)
+	if err != nil {
+		return "", trace.Wrap(err, "Parsing %s", fileUrl)
+	}
+
+	var fetchCmd string
+	var homeDir = fmt.Sprintf("/home/%s", t.sshUser)
```
Comment (r0mant) [inline infra/terraform/file_transfer.go]: Not a big deal, but maybe choose a more reliable location for the file? Theoretically, user's home directory may not be under /home.

File: infra/terraform/file_transfer.go [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+// TransferFile takes file URL which may be S3 or HTTP or local file and transfers it to the machine
+// fileUrl - file to download, could be S3:// or http(s)://
+// command - to run out of installer package
+func (t *terraform) makeRemoteCommand(fileUrl, command string) (string, error) {
+	u, err := url.Parse(fileUrl)
+	if err != nil {
+		return "", trace.Wrap(err, "Parsing %s", fileUrl)
+	}
+
+	var fetchCmd string
```
Comment (r0mant) [inline infra/terraform/file_transfer.go]: I would declare this right before the switch where it's used. In general, I think it's a good idea to declare vars as close to their first usage location as possible.

File: infra/terraform/file_transfer.go [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+		return "", trace.Errorf("Unsupported installer packaging %s", fileUrl)
+	}
+
+	switch u.Scheme {
+	case "s3":
+		if t.Config.AWS == nil {
+			return "", trace.Errorf("AWS config missing, cannot use S3 URLs %s", fileUrl)
+		}
+
+		fetchCmd = fmt.Sprintf(`AWS_ACCESS_KEY_ID=%s \
```
Comment (r0mant) [inline infra/terraform/file_transfer.go]: Is this a reliable approach? Why aren't we using cloud init scripts for this as we used to?

File: infra/terraform/terraform.go [hunk 13]
```diff
@@ -64,45 +71,73 @@ func NewFromState(config Config, stateConfig infra.ProvisionerState) (*terraform
...
+		if err == nil {
+			if withInstaller {
+				nodes := r.pool.Nodes()
+				if len(nodes) == 0 { // should not happen, and doesn't make sense to retry
+					return nil, trace.Errorf("Zero nodes were allocated")
+				}
+				r.installerIP = nodes[0].Addr()
+				return nodes[0], nil
+			}
+			return nil, nil
```
Comment (r0mant) [inline infra/terraform/terraform.go]: Usually, if a function returns `nil` error, the called expects the first value to be usable, so I would avoid returning `nil, nil`. Return an appropriate error instead, like `trace.NotFound()` or `trace.BadParameter()`.

File: infra/terraform/terraform.go [hunk 14]
```diff
@@ -64,45 +71,73 @@ func NewFromState(config Config, stateConfig infra.ProvisionerState) (*terraform
...
+		case <-ctx.Done():
+			return nil, trace.Wrap(err, "Terraform creation timed out")
+		case <-time.After(terraformRepeatAfter):
+			continue
+		}
+	}
+
+}
+
+func (r *terraform) terraform() (err error, retry bool) {
```
Comment (r0mant) [inline infra/terraform/terraform.go]: That's very unusual to have error as a first return argument (unless it's the only one). Can you swap them?

File: lib/system/file.go [hunk 15]
```diff
@@ -63,6 +64,72 @@ func CopyFileWithPerms(dst, src string, perm os.FileMode) error {
...
+	}
+
+	for _, entry := range entries {
+		srcPath := filepath.Join(src, entry.Name())
+		dstPath := filepath.Join(dst, entry.Name())
+
+		if entry.IsDir() {
+			err = copyAll(srcPath, dstPath, fileCount)
+			if err != nil {
+				return err
```
Comment (r0mant) [inline lib/system/file.go]: `trace.Wrap(err)`

File: lib/system/file.go [hunk 16]
```diff
@@ -63,6 +64,72 @@ func CopyFileWithPerms(dst, src string, perm os.FileMode) error {
...
+		}
+
+		if entry.Mode()&os.ModeSymlink != 0 {
+			log.Warning("Symlinks are not copied: %s", srcPath)
+			continue
+		}
+
+		err = CopyFile(srcPath, dstPath)
+		if err != nil {
+			return err
```
Comment (r0mant) [inline lib/system/file.go]: `trace.Wrap(err)`

File: lib/system/file.go [hunk 17]
```diff
@@ -63,6 +64,72 @@ func CopyFileWithPerms(dst, src string, perm os.FileMode) error {
...
+		*fileCount++
+		return CopyFile(src, filepath.Join(dst, filepath.Base(src)))
+	}
+
+	_, err = os.Stat(dst)
+	if err != nil && !os.IsNotExist(err) {
+		return trace.ConvertSystemError(err)
+	}
+
+	if err != nil && os.IsNotExist(err) {
```
Comment (r0mant) [inline lib/system/file.go]: I think in this case it would be clearer to combine this with the above:

```go
if err != nil {
  if !os.IsNotExist(err) {
    return ...
  } 
  err = os.MkdirAll(...)
  ...
}
```

File: lib/system/file.go [hunk 18]
```diff
@@ -63,6 +64,72 @@ func CopyFileWithPerms(dst, src string, perm os.FileMode) error {
...
+
+		if entry.IsDir() {
+			err = copyAll(srcPath, dstPath, fileCount)
+			if err != nil {
+				return err
+			}
+		}
+
+		if entry.Mode()&os.ModeSymlink != 0 {
+			log.Warning("Symlinks are not copied: %s", srcPath)
```
Comment (r0mant) [inline lib/system/file.go]: Warningf?

File: e2e/framework/test_context.go (hunk 2 above)
Comment (DenisMishin) [inline e2e/framework/test_context.go]: https://docs.google.com/document/d/1lwbmjakDVan8haGwHUX65r5OR15tTSiFsz6qKKNVt5k/edit

just like we're getting rid of static variables; to allow concurrent multi-suite tests

File: e2e/framework/test_context.go (hunk 3 above)
Comment (DenisMishin) [inline e2e/framework/test_context.go]: that's pre-existing code; many structs meaning is unclear to me, and especially everything inside this test_context class is subject to massive changes in the coming days;

File: infra/config.go (hunk 6 above)
Comment (DenisMishin) [inline infra/config.go]: the meaning is described at the URL in the struct-level godoc.

File: infra/terraform/file_transfer.go (hunk 11 above)
Comment (DenisMishin) [inline infra/terraform/file_transfer.go]: that whole thing is temporary implementation, see FIXME at top

File: infra/terraform/file_transfer.go (hunk 12 above)
Comment (DenisMishin) [inline infra/terraform/file_transfer.go]: as per aws cli docs; also it's not necessarily an AWS instance, and in general I don't like global inits which is all over robotest code. 

once that'll be replaced with more fine-grained SSH, it'll help to shield sensitive key stuff from logs.

Comment (DenisMishin) [inline infra/terraform/file_transfer.go]: also cloud init need be kept at bare minimum, no test logic like fetching installers should live there.

File: lib/system/file.go [hunk 19]
```diff
@@ -63,6 +64,72 @@ func CopyFileWithPerms(dst, src string, perm os.FileMode) error {
 	return nil
 }
 
+// Recursively copy
+// dst must always be a directory
+// src may be either a dir or a file
+func CopyAll(src, dst string) (err error, fileCount uint) {
```
Comment (a-palchikov) [inline lib/system/file.go]: Please return errors last

File: lib/system/file.go [hunk 20]
```diff
@@ -63,6 +64,72 @@ func CopyFileWithPerms(dst, src string, perm os.FileMode) error {
...
+	fileCount = 0
+	err = copyAll(src, dst, &fileCount)
+	return err, fileCount
+}
+
+func copyAll(src, dst string, fileCount *uint) (err error) {
+	src = filepath.Clean(src)
+	dst = filepath.Clean(dst)
+
+	log.Debugf("CopyAll %s %s", src, dst)
```
Comment (a-palchikov) [inline lib/system/file.go]: `%s -> %s` to define the operation direction

File: lib/system/file.go [hunk 21]
```diff
@@ -63,6 +64,72 @@ func CopyFileWithPerms(dst, src string, perm os.FileMode) error {
...
+
+	for _, entry := range entries {
+		srcPath := filepath.Join(src, entry.Name())
+		dstPath := filepath.Join(dst, entry.Name())
+
+		if entry.IsDir() {
+			err = copyAll(srcPath, dstPath, fileCount)
+			if err != nil {
+				return err
+			}
```
Comment (a-palchikov) [inline lib/system/file.go]: You need to `continue` here - otherwise the function will try to `CopyFile` a directory.
On a side note - why reinvent the [`filepath.Walk`](https://golang.org/pkg/path/filepath/#Walk)?

File: assets/terraform/aws/bootstrap.sh [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+mount /var/lib/gravity
+mount /var/lib/data
+mkdir -p /var/lib/gravity/planet/etcd
+mount /var/lib/gravity/planet/etcd
+
+chown -R 1000:1000 /var/lib/gravity /var/lib/data /var/lib/gravity/planet/etcd
+sed -i.bak 's/Defaults    requiretty/#Defaults    requiretty/g' /etc/sudoers
+
+# robotest might SSH before bootstrap script is complete (and will fail)
+touch /var/lib/bootstrap_complete
```
Comment (DenisMishin) [inline assets/terraform/aws/bootstrap.sh]: its just tested for existence

File: infra/terraform/file_transfer.go [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+	case "https":
+		fetchCmd = fmt.Sprintf("wget %s -O %s/", fileUrl, outFile)
+	case "":
+	case "gs":
+	default:
+		// TODO : implement SCP and GCLOUD methods
+		return "", fmt.Errorf("Unsupported URL schema %s", fileUrl)
+	}
+
+	cmd := fmt.Sprintf(`echo Testing if bootstrap completed && test -f /var/lib/bootstrap_complete && \
```
Comment (DenisMishin) [inline infra/terraform/file_transfer.go]: that's temporary implementation, to be replaced.

**Comment by sofuture** [comment]
What does 1. mean? Can't use TF defaults?

File: assets/terraform/aws/bootstrap.sh [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+apt install -y python-pip lvm2
+pip install --upgrade awscli
+
+set -euo pipefail
+
+pvcreate -ff /dev/xvde
+vgcreate data /dev/xvde
+lvcreate -n var_lib_gravity -l 20%VG data
+lvcreate -n var_lib_data -l 20%VG data
+lvcreate -n etcd -l 60%VG data
```
Comment (sofuture) [inline assets/terraform/aws/bootstrap.sh]: In practice this needs a real 'physical' device of it's own. Instance storage if the instance has it, or a 50GB 2500IOPS io1 SSD on AWS.

Comment (DenisMishin) [inline assets/terraform/aws/bootstrap.sh]: etcd or each ?

Comment (DenisMishin) [inline assets/terraform/aws/bootstrap.sh]: >> Can't use TF defaults?
Yes, TF would be just provisioning templates; all configuration values would come top-down.

File: assets/terraform/aws/bootstrap.sh [hunk 25]
```diff
@@ -0,0 +1,36 @@
+#!/bin/bash
```
Comment (sofuture) [inline assets/terraform/aws/bootstrap.sh]: Once we are done bootstrapping, we should disable internet egress with iptables. We've shipped regressions of things that needing internet before, without doing that. If we find exceptions that we want to allow in CI, like say NTP, can enable those port by port.

See: https://github.com/gravitational/anypoint/blob/develop/terraform/onprem.tf#L454

File: assets/terraform/aws/network.tf [hunk 26]
```diff
@@ -0,0 +1,44 @@
...
+
+    # install wizard
+    ingress {
+        from_port   = 61009
+        to_port     = 61009
+        protocol    = "tcp"
+        cidr_blocks = ["0.0.0.0/0"]
+    }
+
+    ingress {
```
Comment (sofuture) [inline assets/terraform/aws/network.tf]: Out of ultra caution, we should not allow blanket intra-node communcation -- and should make it port-by-port. Luckily I have a relatively complete example of this :)

```
    # installer ports
    ingress {
        from_port = 61008
        to_port = 61010
        protocol = "tcp"
        self = true
    }

    # installer ports
    ingress {
        from_port = 61022
        to_port = 61024
        protocol = "tcp"
        self = true
    }

    # bandwidth checker (for pre-checks)
    ingress {
        from_port   = 4242
        to_port     = 4242
        protocol    = "tcp"
        cidr_blocks = ["0.0.0.0/0"]
    }

    # k8s health check
    ingress {
        from_port = 10250
        to_port = 10250
        protocol = "tcp"
        cidr_blocks = ["0.0.0.0/0"]
    }

    # serf peer-to-peer
    ingress {
        from_port = 7496
        to_port = 7496
        protocol = "tcp"
        self = true
    }

    # kubernetes apiserver insecure
    ingress {
        from_port = 8080
        to_port = 8080
        protocol = "tcp"
        self = true
    }

    # kubernetes apiserver secure
    ingress {
        from_port = 6443
        to_port = 6443
        protocol = "tcp"
        self = true
    }

    # etcd peer-to-peer
    ingress {
        from_port = 2380
        to_port = 2380
        protocol = "tcp"
        self = true
    }

    # etcd client
    ingress {
        from_port = 2379
        to_port = 2379
        protocol = "tcp"
        self = true
    }

    # etcd legacy client
    ingress {
        from_port = 4001
        to_port = 4001
        protocol = "tcp"
        self = true
    }

    # etcd legacy peer-to-peer
    ingress {
        from_port = 7001
        to_port = 7001
        protocol = "tcp"
        self = true
    }

    # docker registry
    ingress {
        from_port = 5000
        to_port = 5000
        protocol = "tcp"
        self = true
    }

    # overlay networking
    ingress {
        from_port = 8472
        to_port = 8472
        protocol = "udp"
        self = true
    }

    # gravity profiling port
    ingress {
        from_port = 6060
        to_port = 6060
        protocol = "tcp"
        self = true
    }

    # teleport services (SSH server, proxy service, proxy tunnel, auth service)
    ingress {
        from_port = 3022
        to_port = 3025
        protocol = "tcp"
        self = true
    }

    # teleport web
    ingress {
        from_port = 3080
        to_port = 3080
        protocol = "tcp"
        self = true
    }

    # internal k8s services
    ingress {
        from_port = 30000
        to_port = 32767
        protocol = "tcp"
        self = true
    }

    # kubernetes kubelet
    ingress {
        from_port = 10248
        to_port = 10249
        protocol = "tcp"
        self = true
    }

    # kubernetes kubelet
    ingress {
        from_port = 10255
        to_port = 10255
        protocol = "tcp"
        self = true
    }

    # gravity services (ops service, pack service, health test)
    ingress {
        from_port = 3008
        to_port = 3010
        protocol = "tcp"
        self = true
    }

    # planet agent RPC
    ingress {
        from_port = 7575
        to_port = 7575
        protocol = "tcp"
        self = true
    }

    # serf RPC
    ingress {
        from_port = 7373
        to_port = 7373
        protocol = "tcp"
        self = true
    }
```

File: assets/terraform/aws/node.tf [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+    root_block_device {
+        delete_on_termination = true
+        volume_type = "io1"
+        volume_size = "50"
+        iops = 500
+    }
+
+    # etc and other data device
+    ebs_block_device = {
+        volume_size = "500"
```
Comment (sofuture) [inline assets/terraform/aws/node.tf]: We can make this way smaller: 100gb standard SSD should be fine (split etcd into it's own 50gb 2500IOPS io1 though)

File: assets/terraform/aws/node.tf [hunk 28]
```diff
@@ -0,0 +1,43 @@
...
+        volume_type = "io1"
+        device_name = "/dev/xvde"
+        iops = 3000
+        delete_on_termination = true
+    }
+
+    # gravity/docker data device
+    ebs_block_device = {
```
Comment (sofuture) [inline assets/terraform/aws/node.tf]: This is unused in bootstrap, I assume it's a device for docker device mapper. It can be made much smaller and lower performance as well: 100GB, 2000IOPS is plenty

File: assets/terraform/azure/bootstrap.sh [hunk 29]
```diff
@@ -0,0 +1,36 @@
...
+apt install -y python-pip lvm2
+pip install --upgrade awscli
+
+set -euo pipefail
+
+pvcreate -ff /dev/sdc
+vgcreate data /dev/sdc
+lvcreate -n var_lib_gravity -l 20%VG data
+lvcreate -n var_lib_data -l 20%VG data
+lvcreate -n etcd -l 30%VG data
```
Comment (sofuture) [inline assets/terraform/azure/bootstrap.sh]: same issue with etcd -- I know disks on azure are pretty different (worse), so maybe doesn't apply, but I would assume it does.

File: assets/terraform/azure/bootstrap.sh [hunk 30]
```diff
@@ -0,0 +1,36 @@
+#!/bin/bash
```
Comment (sofuture) [inline assets/terraform/azure/bootstrap.sh]: ditto with egress

File: assets/terraform/azure/network.tf [hunk 31]
```diff
@@ -0,0 +1,68 @@
...
+    priority                   = 1020
+    direction                  = "Inbound"
+    access                     = "Allow"
+    protocol                   = "TCP"
+    source_port_range          = "*"
+    destination_port_range     = 61009
+    source_address_prefix      = "*"
+    destination_address_prefix = "*"
+  }
+}
```
Comment (sofuture) [inline assets/terraform/azure/network.tf]: ditto with whitelisting intra-cluster ports

File: e2e/framework/test_context.go [hunk 32]
```diff
@@ -375,7 +351,22 @@ func newContextConfig(configFile string) error {
...
+	}
+
+	err = validator.New().Struct(TestContext)
+	if err == nil {
+		return nil
+	}
+
+	log.Errorf("Configuration file %s has errors", configFile)
+	for _, fieldError := range err.(validator.ValidationErrors) {
+		log.Errorf("   Field %s=%v fails rule %s", fieldError.Field(), fieldError.Value(), fieldError.Tag())
```
Comment (sofuture) [inline e2e/framework/test_context.go]: Thank you for this :+1:

File: infra/config.go (hunk 6 above)
Comment (sofuture) [inline infra/config.go]: Still, it makes for nicer autogenerated docs to follow the idiom of 'FooBar ...' comments on every exported `FooBar`

File: infra/terraform/file_transfer.go (hunk 12 above)
Comment (sofuture) [inline infra/terraform/file_transfer.go]: I agree, fetching installers should be separate from provisioning infrastructure -- it's a wart in how customer uses this right now, they have a `while aws sleep` loop in cloud-init to get installer. This starts to give us a little more control over flow.

File: assets/terraform/aws/bootstrap.sh (hunk 24 above)
Comment (sofuture) [inline assets/terraform/aws/bootstrap.sh]: Just etcd -- and updated figures: 30gb 1500IOPS.

File: e2e/framework/framework.go [hunk 33]
```diff
@@ -298,7 +298,7 @@ func fetchOpsCenterLogs() error {
 		return trace.Wrap(err, "failed to collect site report")
 	}
 
-	err = system.CopyFile(output, tempOutput.Name())
+	err = system.CopyFile(tempOutput.Name(), output)
```
Comment (DenisMishin) [inline e2e/framework/framework.go]: stdlib is counter intuitive by itself... when you do cp its always source, then destination

File: assets/terraform/aws/bootstrap.sh (hunk 25 above)
Comment (DenisMishin) [inline assets/terraform/aws/bootstrap.sh]: I'll make all TF template enhancements part of a separate PR after I test all deployment params such as different distros, etc; don't hold this specific PR for too long.

File: infra/terraform/terraform.go [hunk 34]
```diff
@@ -64,45 +71,73 @@ func NewFromState(config Config, stateConfig infra.ProvisionerState) (*terraform
 }
 
 func (r *terraform) Create(withInstaller bool) (installer infra.Node, err error) {
-	err = system.CopyFile(filepath.Join(r.stateDir, "terraform.tf"), r.ScriptPath)
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
+	ctx, cancel := context.WithTimeout(context.Background(), terraformTimeout)
```
Comment (DenisMishin) [inline infra/terraform/terraform.go]: that would be a chain reaction; it'll have to go top-down between most functions in infra, framework and lib package; which is fine but it should be subject to separate PR, likely coupled with changes to more fine-grained SSH command execution. 

please approve as-is.

Comment (a-palchikov) [inline infra/terraform/terraform.go]: Ok
